### PR TITLE
Car 44 clustering, most liquid ads, docker configuring

### PR DESCRIPTION
**Pull request by BobriK502**

### CODE REVIEW

File: src/config/config.ts [hunk 1]
```diff
@@ -8,9 +8,9 @@ export const limitForSavedFilters = 2;
 export const mail = {
   auth: {
     pass: process.env.MAIL_PASS,
-    user: process.env.MAIL_USER
+    user: process.env.MAIL_USER,
   },
-  service: 'Gmail'
+  service: 'Gmail',
```
Comment (AntonNarkevich) [inline src/config/config.ts]: No-no :) we don't need a comma here

File: src/config/config.ts [hunk 2]
```diff
@@ -30,7 +30,7 @@ export const codeErrors = {
   MONGO_DUPLICATE_ERROR: 11000,
   ONLINER_PARSE_ERROR: 106,
   REQUIRED_FIELD: 110,
-  VALIDATION_ERROR: 105
+  VALIDATION_ERROR: 105,
```
Comment (AntonNarkevich) [inline src/config/config.ts]: ,?

File: src/config/config.ts [hunk 3]
```diff
@@ -51,7 +51,7 @@ export const emailActions = {
 export const awsConfig = {
   accessKeyId: process.env.AWS_ACCESS_KEY,
   region: process.env.AWS_REGION,
-  secretAccessKey: process.env.AWS_SECRET_KEY
+  secretAccessKey: process.env.AWS_SECRET_KEY,
```
Comment (AntonNarkevich) [inline src/config/config.ts]: ,?

File: src/controllers/adController.ts [hunk 4]
```diff
@@ -30,6 +31,15 @@ export const getAds = async (ctx: Koa.Context) => {
   }
 };
 
+export const getMostLiquid = async (ctx: Koa.Context) => {
+  try {
+    const mostLiquid = await getMostLiquidAds();
+    ctx.body = mostLiquid;
+  } catch (e) {
+    ctx.body = e;
```
Comment (AntonNarkevich) [inline src/controllers/adController.ts]: Should we set a status code here?

File: src/index.ts [hunk 5]
```diff
@@ -35,18 +43,34 @@ const parse = schedule.scheduleJob(triggerSchedule, async () => {
     torTriggerer.close();
   }
   await https.get('https://hchk.io/c12a23b6-276d-4269-9316-d3353af47052');
+  await calculateAllLiquidity();
 });
 
-mongoose.connect(db, { useMongoClient: true });
-mongoose.set('debug', true);
+  cluster.on('online', (worker) => {
+    global.console.log('Worker ' + worker.process.pid + ' is online');
```
Comment (AntonNarkevich) [inline src/index.ts]: `` here :)
well, not critical, agree

File: src/repositories/tempAdRepository.ts [hunk 6]
```diff
@@ -28,6 +28,7 @@ export const updateAds = async () => {
             sourceName: doc.sourceName,
             sourceUrl: doc.sourceUrl,
             year: doc.year
+
```
Comment (AntonNarkevich) [inline src/repositories/tempAdRepository.ts]: What a crazy empty line? :)

**Review by AntonNarkevich — CHANGES_REQUESTED** [review]
Looks really good. Need to fix some minors and let's merge.

File: src/config/config.ts (hunk 1 above)
Comment (BobriK502) [reply src/config/config.ts]: Sorry, it's webstorm autoformater I forget about it =)

File: src/index.ts [hunk 7]
```diff
@@ -12,18 +12,26 @@ import * as mongoose from 'mongoose';
...
 import { Api } from './parsers';
 
-const server = new Koa();
+
+import * as cluster from 'cluster';
+import * as os from 'os';
+
+if (cluster.isMaster) {
+  const cpuCount = os.cpus().length;
+  for (let pInd = 0; pInd < cpuCount; pInd += 1) {
```
Comment (AntonNarkevich) [inline src/index.ts]: += 👍

File: src/index.ts [hunk 8]
```diff
@@ -35,18 +43,34 @@ const parse = schedule.scheduleJob(triggerSchedule, async () => {
     torTriggerer.close();
   }
   await https.get('https://hchk.io/c12a23b6-276d-4269-9316-d3353af47052');
```
Comment (AntonNarkevich) [inline src/index.ts]: Why is this hardcoded?

File: src/index.ts [hunk 9]
```diff
@@ -35,18 +43,34 @@ const parse = schedule.scheduleJob(triggerSchedule, async () => {
...
+  await calculateAllLiquidity();
 });
 
-mongoose.connect(db, { useMongoClient: true });
-mongoose.set('debug', true);
+  cluster.on('online', (worker) => {
+    global.console.log(`Worker   ${worker.process.pid}  is online`);
```
Comment (AntonNarkevich) [inline src/index.ts]: Wait, why are we doing this? Why global? Why are we using console at all. We should switch to winston.

File: src/index.ts [hunk 10]
```diff
@@ -35,18 +43,34 @@ const parse = schedule.scheduleJob(triggerSchedule, async () => {
...
+
+} else {
+  const server = new Koa();
+
+  mongoose.connect(db, { useMongoClient: true });
+  mongoose.connect(db, { useMongoClient: true });
+  mongoose.set('debug', true);
 
-(mongoose as any).Promise = bluebird;
+  (mongoose as any).Promise = bluebird;
```
Comment (AntonNarkevich) [inline src/index.ts:64]: Had to do the same at my current project lol

File: src/repositories/adRepository.ts [hunk 11]
```diff
@@ -103,7 +115,7 @@ export const get = async (
   sort?: any
 ): Promise<IAdModel[]> => {
   return (await Ad.find(filter || {})
-    .limit(limit || 20)
+    .limit(limit === null ? null : limit || 20)
```
Comment (AntonNarkevich) [inline src/repositories/adRepository.ts]: This line looks not clear for me. Strange ternary combined with ||. Please refactor. Possibly extract as a variable.

File: src/routes/ad.ts [hunk 12]
```diff
@@ -1,11 +1,13 @@
...
-import { getAds, getLiquidity, getSavedSearchAds } from '../controllers/adController';
+
+import { getAds, getLiquidity, getMostLiquid,getSavedSearchAds } from '../controllers/adController';
 
 const router = new Router();
 
 router.post('/', getAds);
 router.get('/saved', getSavedSearchAds);
+router.get('/most_liquid', getMostLiquid);
 router.post('/get-liquidity', getLiquidity);
```
Comment (AntonNarkevich) [inline src/routes/ad.ts]: Please don't use minus sign in urls. This is a moveton. Use lodashes insted as you did above. Please refactor.

File: src/services/adService.ts [hunk 13]
```diff
@@ -2,14 +2,54 @@ import * as async from 'async';
...
+};
+
+export const getAdByModelId = async (modelId: string) => {
+  return await getByModelId(modelId);
+};
+
+export const getSoldAd = async (
+  modelsBuffer: string[],
+  bodyTypeIdsBuffer: string[]
+): Promise<any> => {
```
Comment (AntonNarkevich) [inline src/services/adService.ts]: Why is it any, not boolean here?

File: src/services/index.ts [hunk 14]
```diff
@@ -207,6 +239,40 @@ export const getSavedFiltersAds = async (user: IUser): Promise<ISavedFilterAds[]
...
+  const bodyTypes = await getAllBodyTypes();
+  const models = await getModels();
+  const totalSold = await AdService.countSoldAds();
+
+  for (const model of models) {
+    for (const bodyType of bodyTypes) {
+      const totalSoldInConfig = await AdService.countSoldWithFilter(model.id, bodyType.id);
+      const ads = await getAds({ modelId: [model.id], bodyTypeId: [bodyType.id] }, null);
+      const adPrices = ads.map((item: any) => item.price);
+      adPrices.sort((item: any, nextItem: any) => {
```
Comment (AntonNarkevich) [inline src/services/index.ts]: Lodash has a method not to do this like that. You should somehow specify a path to a property. Or do the arrow function returning a value to sort by.

**Review by AntonNarkevich — CHANGES_REQUESTED** [review]
Please fix some minors and do some refactoring. Thanks.

File: src/index.ts (hunk 9 above)
Comment (BobriK502) [reply src/index.ts]: Fixed.Now used winston 2.4 with custom timestamp for logging.It seems like this =)
logger

File: src/index.ts (hunk 8 above)
Comment (BobriK502) [reply src/index.ts]: Fixed

File: src/services/adService.ts (hunk 13 above)
Comment (BobriK502) [reply src/services/adService.ts]: It should have <IAd> type `cause this method returns all sold ads with current mark and body type.Fixed

**Comment by AntonNarkevich** [comment]
Please fix the tests

**Comment by AntonNarkevich** [comment]
Finally, we reached the 75% coverage. My congrats! @BobriK502 @ElizavetaBuraya @StankAnt @AnatolyBelobrovik 
Great job, guys!

**Pull request merged**